### PR TITLE
Issue 319 - bouncer worker improvement

### PR DESCRIPTION
**Pull request by carmenfan**
- added a timeout error
- added a timeout on all process (default is 3hrs, configuration on the config file)
- use amqp promise api instead of callback
- auto retry when amqp connection drops (default is 3 tries, configuration on the config file)
- add an error check on toy project - if settings isn't found it should reject the promise.

### CODE REVIEW

File: tools/bouncer_worker/bouncer_worker.js [hunk 1]
```diff
@@ -270,117 +294,110 @@
...
+						value: code,
+						database: cmdDatabase,
+						project: cmdProject,
+						user
+					}, true);
 				}
 			}
 
+		}).catch((code) => {
+			const err =  code? code : ERRCODE_BOUNCER_CRASH;
```
Comment (sanmont3drepo) [inline tools/bouncer_worker/bouncer_worker.js]: An alternate form : `const err = code || ERRCODE_BOUNCER_CRASH`

File: tools/bouncer_worker/bouncer_worker.js [hunk 2]
```diff
@@ -270,117 +294,110 @@
...
+			callback({
+				value: err,
+				database: cmdDatabase,
+				project: cmdProject,
+				user
+			}, true);
+			logger.info(`[FAILED] Executed command: ${command} ${cmdParams.join(" ")}`, err);
```
Comment (sanmont3drepo) [inline tools/bouncer_worker/bouncer_worker.js]: `logger.error`

File: tools/bouncer_worker/bouncer_worker.js [hunk 3]
```diff
@@ -425,19 +442,47 @@
...
-			else
-			{
-				conn.createChannel(function(err, ch){
-					ch.assertQueue(conf.rabbitmq.callback_queue, { durable: true });
-					listenToQueue(ch, conf.rabbitmq.worker_queue, conf.rabbitmq.task_prefetch || 4);
-					listenToQueue(ch, conf.rabbitmq.model_queue, conf.rabbitmq.model_prefetch || 1);
+		amqp.connect(conf.rabbitmq.host).then((conn) => {
+			retry = 0;
+			connClosed = false;
+			logger.error("[AMQP] Connected! Creating channel...");
```
[the remark this answers is not in the document]
Comment (carmenfan) [reply tools/bouncer_worker/bouncer_worker.js]: yup...

File: tools/bouncer_worker/bouncer_worker.js [hunk 4]
```diff
@@ -425,19 +442,47 @@
...
+						connectQ();
+					} else {
+						logger.error("[AMQP] Retries exhausted");
+						process.exit(-1);
+					}
+				}
+			});
+
+			conn.on("error", (err)	=> {
+				logger.error("[AMQP] connection error: " + err.message);
```
[the remark this answers is not in the document]
Comment (carmenfan) [reply tools/bouncer_worker/bouncer_worker.js]: It is. 2 reasons.
- we don't want to retry to connect if there's an error on the queue connection, it's not necessarily means we have been disconnected
- This gets called way after the the promise would've been resolved, not during the connection is trying to be established

File: tools/bouncer_worker/bouncer_worker.js [hunk 5]
```diff
@@ -425,19 +442,47 @@
...
+					connClosed = true;
+					const maxRetries = conf.rabbitmq.maxRetries || 300;
+					logger.error("[AMQP] connection closed.");
+					if(++retry <= maxRetries) {
+						logger.error(`[AMQP] Trying to reconnect [${retry}/${maxRetries}]...`);
+						connectQ();
+					} else {
+						logger.error("[AMQP] Retries exhausted");
+						process.exit(-1);
+					}
```
[the remark this answers is not in the document]
Comment (carmenfan) [reply tools/bouncer_worker/bouncer_worker.js]: yup

**Pull request merged**